all/generate_gateway.sh

```diff
@@ -0,0 +1,300 @@
+mkdir -p $TMPDIR/src/proto/
+cp $FILE $TMPDIR/src/proto/
+
+INCLUDES_ARG=""
+if [[ ! -z "$EXTRA_INCLUDES" ]]; then
+  INCLUDES_ARG="-i $EXTRA_INCLUDES"
+fi
+docker run -v $TMPDIR/src:/defs namely/protoc-all -f proto/`basename $FILE` -l go --with-gateway $INCLUDES_ARG
+
+echo "Generating entry-point code."
+mkdir -p $TMPDIR/src/pkg/main/
+
+cat << ENTRYPOINT >> $TMPDIR/src/pkg/main/main.go
+// AUTOMATICALLY GENERATED BY namely/docker-protoc
+// DO NOT EDIT. DO NOT CHECK INTO GITHUB. DO NOT PASS GO. DO NOT COLLECT \$200.
+package main
+
+import (
+  "fmt"
+  "log"
+  "net/http"
+  "os"
+  "os/signal"
+  "strings"
+  "time"
+
+  "github.com/grpc-ecosystem/grpc-gateway/runtime"
+  "github.com/spf13/viper"
+  "golang.org/x/net/context"
+  "google.golang.org/grpc"
+
+  gw "gen/pb-go/proto"
+)
+
+type proxyConfig struct {
+  host    string
+  port    int
+  swagger string
+}
+
+func SetupMux(ctx context.Context, cfg proxyConfig) *http.ServeMux {
+  mux := http.NewServeMux()
+
+  mux.HandleFunc("/swagger.json", func(w http.ResponseWriter, r *http.Request) {
+    http.ServeFile(w, r, cfg.swagger)
+  })
+
+  opts := []grpc.DialOption{grpc.WithInsecure()}
+  gwmux := runtime.NewServeMux()
+  endpoint := fmt.Sprintf("%v:%v", cfg.host, cfg.port)
+  err := gw.Register${SERVICE}HandlerFromEndpoint(ctx, gwmux, endpoint, opts)
+  if err != nil {
+    log.Fatalf("Could not register gateway: %v", err)
+  }
+  mux.Handle("/", gwmux)
+
+  return mux
+}
+
+// SetupViper returns a viper configuration object
+func SetupViper() *viper.Viper {
+  viper.SetConfigName("config")
+  viper.AddConfigPath(".")
+  viper.SetEnvPrefix("${SERVICE}")
+  viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+  viper.AutomaticEnv()
+
+  err := viper.ReadInConfig()
+  if err != nil {
+    log.Fatalf("Could not read config: %v", err)
+  }
+
+  return viper.GetViper()
+}
+
+// SignalRunner runs a runner function until an interrupt signal is received, at which point it
+// will call stopper.
+func SignalRunner(runner, stopper func()) {
+  signals := make(chan os.Signal, 1)
+  signal.Notify(signals, os.Interrupt, os.Kill)
+
+  go func() {
+    runner()
+  }()
+
+  fmt.Println("hit Ctrl-C to shutdown")
+  select {
+  case <-signals:
+    stopper()
+  }
+}
+
+func main() {
+
+  cfg := SetupViper()
+  ctx := context.Background()
+  ctx, cancel := context.WithCancel(ctx)
+  defer cancel()
+
+  mux := SetupMux(ctx, proxyConfig{
+    host:    cfg.GetString("grpc_service_host"),
+    port:    cfg.GetInt("grpc_service_port"),
+    swagger: cfg.GetString("swagger.file"),
+  })
+
+  addr := fmt.Sprintf(":%v", cfg.GetInt("proxy.port"))
+  server := &http.Server{Addr: addr, Handler: mux}
+
+  SignalRunner(
+    func() {
+      fmt.Printf("launching http server on %v\n", server.Addr)
+      if err := server.ListenAndServe(); err != nil {
+        log.Fatalf("Could not start http server: %v", err)
+      }
+    },
+    func() {
+      shutdown, _ := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
Is there a reason you're not getting the CancelFunc from this and defer'ing it being called?

Reply:
There's no use for the cancel func in any scope, is there?